models/image.py

```diff
@@ -1242,6 +1105,287 @@ def load(self):
                 if not ( gotim and gotweight and gotflags ):
                     raise FileNotFoundError( "Failed to load at least one of image, weight, flags" )
 
+    def get_upstream_provenances(self):
+        """Collect the provenances for all upstream objects.
+
+        This does not recursively go back to the upstreams of the upstreams.
+        It gets only the provenances of the immediate upstream objects.
+
+        Provenances that are the same (have the same hash) are combined (returned only once).
+
+        This is what would generally be put into a new provenance's upstreams list.
+
+        Note that upstream_images must each have the other related products
+        like sources, psf, wcs, etc. already loaded.
+        This happens when the objects are used to produce, e.g., a coadd or
+        a subtraction image, but they would not necessarily be loaded automatically from the DB.
+        To load those products (assuming all were previously committed with their own provenances)
+        use the load_upstream_products() method.
+
+        IMPORTANT RESTRICTION: to maintain the ability of a downstream to recover its upstreams
+        using the provenance (which is the definition of why we need a provenance) it is not
+        allowed for images with the same provenance to have related products (e.g., a SourceList)
+        that have different provenances.  This is because the downstream would not know which
+        SourceList to use.  Images from different instruments, or a coadded reference vs.
+        a new image, would have different provenances, so their products could (and indeed must)
+        have different provenances. But images from the same instrument with the same provenance
+        should all be produced using the same code and parameters, otherwise it will be impossible
+        to know which product was processed in which way.
```

Comment:
...I'm not sure this restirction is a good idea.  If we do have this restriction, then a lot of the provenance stuff we have in place isn't actually necessary. Once you've got the image id and the code version, the provence is uniquely defined for all downstream products.

This restriction will make it impossible for us to do tests by trying different paremeters for downstream processes (catalogs, zeropoints, whatever) off of the same image.  This is something I suspect we're going to want to have to do.  (In the past, I've done it ad hoc, which of course becomes a pain to keep track of, and difficult to do in bulk.  The provenace system you've defined alows us to keep track of this, which is great, but adding this restriction totally undermines the flexibility of it.

If we DO go this way with this restriction, then I'd strongly recommend we think about other places where we can simplify the code, because I think we have a lot of complexity that becomes unnecessary if this restriction is in place.

I think there's still a fundamental design problem here.  To make a coadd, you need to have the source list, and perhaps the WCS, of the images that go into this coadd.  Because right now for a coadd, you list the images themselves as the upstreams, you now are introducing this requirement that there can only be one downstream of any time for an image.  But I don't think that's the right solution.

The right solution is, I think the one I proposed a couple of months ago.  List the things that actually define what goes into a coadd, even if it's not entirely intuitive. For sky images, that could be a zeropoint object, as a zeropoint will point uniquely to upstream WCS, sourcelist, and image.  That doesn't work for flatfields, though, which won't have any of those, only images.  Maybe the solution is an ImageProducts table, but as I said when we were talking about that, if we add that, then we should also remove the direct foreign keys from things like source lists, WCS, etc., so that the imageproducts table becomes the way to find those linkages.  (What I don't want is to have the linkages multiply defined.)

Reply:
I think the docstring is not clear. If image has provenance A, then it can have SourceList provenance 1 and 2 associated with it, just not *relative to this specific coadd*. I will have to fix the text to be more clear. 

So the way it should be: 
Any Image can have any number of different products with different provenances. 
If you put two images with the same provenance *into a coadd* then the associated products (e.g., SourceList) that you pick to use along with those upstream images, must share a provenance as well. They could have other SourceLists but those are not involved in making this specific coadd image.

Reply:
Oh!  OK, that's not nearly as bad as what I was thinking... in fact, it's not bad at all.  Yeah, the docstring sounds like things are far more restrictd than they really are.  But, yeah, makes sense that all the things going into a coadd need to have the same provenance, and that that provenance is what contributes to the coadd's provenance.

OK, I will keep reading further down on the PR.

Reply:
The restriction is really about what's loaded into the the associated fields in self.upstream_images (which, as documented,must already be loaded for this routine to work).